src/lib/device/virtual_dev.cpp

```diff
@@ -197,7 +197,7 @@ BlkAllocStatus VirtualDev::do_alloc_blk(blk_count_t nblks, const blk_alloc_hints
         size_t attempt{0};
 
         do {
-            chunk = m_chunk_selector->select(nblks, hints);
+            chunk = m_chunk_selector->select_chunk(nblks, hints).get();
```

Comment:
@hkadayam This is dangerous as raw ptr/shared ptr are mixed used.

The cshare<chunk>  was born in `PhysicalDev::create_chunks` and all the way pass to chunkSelector, however the consumer part(allocator, etc) use raw ptr get from the cshare<chunk>.

It is functioning fine as a reference to every chunk was kept at `m_chunks_map` but it is a bad practice. Maybe lets have a ticket to move those *chunk to cshare<chunk> ?

Reply:
I don't agree with this change of doing a .get() and select_chunk() returning a shared_ptr. Why incrementing atomic reference when all you are doing is accessing an in-memory member. I am not asking to change now, but will change in next PR.

I would rather return a chunk_id and look it up on the all chunks in vdev. If vdev doesn't maintain all chunks, then I would rather ask it to return a reference to chunks. This is a critical path for homestore, don't want unneccessary atomic reference increment and decrements.

Reply:
the upstream (PhysicalDev::create_chunks ) use shared_ptr<chunk>, if we dont want ref counting we can change the interface of add/select to ref/raw_ptr.  It could work as pdev holds the shared_ptr that ensuring the chunks will not be de-construct,  just not safe.